### PR TITLE
init: use std::thread for ThreadImport()

### PR DESCRIPTION
**Pull request by fanquake**
[Mentioned](https://github.com/bitcoin/bitcoin/pull/19142#issuecomment-638090759) in #19142, which removed the `boost::interruption_point()`
in `ThreadImport()`.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

File: src/init.cpp [hunk 1]
```diff
@@ -1853,7 +1852,8 @@ bool AppInitMain(const util::Ref& context, NodeContext& node)
         vImportFiles.push_back(strFile);
     }
 
-    threadGroup.create_thread([=, &chainman] { ThreadImport(chainman, vImportFiles); });
+    std::thread import(&TraceThread<std::function<void()>>, "loadblk", [=, &chainman]{ ThreadImport(chainman, vImportFiles); });
+    import.detach();
```
Comment (hebasto) [inline src/init.cpp:1856]: nit: The `AppInitMain()` is a way large function. Could the scope of the `import` variable be limited by block?

**Review by hebasto — APPROVED** [review]
ACK aeb100f79ec6ff816ee26344b4eef70001dd51a6, tested on Linux Mint 19.3 (x86_64) with `-reindex` command-line option and thread monitoring via htop.

**Comment by hebasto** [comment]
Keep testing and the second instance of "loadblk" thread appears:

Screenshot from 2020-06-07 10-02-35

Is it ok?

**Comment by MarcoFalke** [comment]
why is is ok to not wait for the thread to exit (join) before shutdown?

**Comment by sipa** [comment]
@MarcoFalke AFAIK, there is no problem with that. The application won't quit until all its threads have terminated.

**Comment by MarcoFalke** [comment]
What I mean is that during shutdown we assume the chain clients don't receive (let's say) block connected signals anymore because they are already flushed. After this change, it is (theoretically, I haven't checked) possible that ThreadImport generates events that will be sent to objects that can't handle them or that ThreadImport reads from objects that have been deleted.

Am I missing something obvious?

**Comment by donaloconnor** [comment]
> @MarcoFalke AFAIK, there is no problem with that. The application won't quit until all its threads have terminated.

I thought that once main returns, any detached threads will not be waited on and stack unwinding does not occur risking danger of uncalled dtors.

I tried a sample here and it seems to be the case?
```
#include <thread>
#include <chrono>
#include <iostream>

using namespace std::chrono_literals;
struct Destruct
{
    ~Destruct() { std::cout << "Dtor called" << std::endl; }
};

void ThreadProc()
{
  Destruct d;
  std::this_thread::sleep_for(1s);
}

int main(int argc, char* argv[])
{
  std::cout << "Hello" << std::endl;

  std::thread t(ThreadProc);
  //t.join();
  t.detach();

  std::cout << "Bye" << std::endl;
  return 0;
}
```

Output is Hello,Bye instead of Hello,Dtor called,Bye.

**Comment by sipa** [comment]
@donaloconnor I would certainly have expected the code above to sleep 1 s before exiting, so it does seem I'm wrong.

@MarcoFalke That's a good point. It seems we'll need to keep the std::thread object around somehow, and join it in the right order at shutdown.

**Comment by laanwj** [comment]
> @MarcoFalke That's a good point. It seems we'll need to keep the std::thread object around somehow, and join it in the right order at shutdown.

Yes. It needs to be join-ed on shutdown to make sure it has terminated before tearing down validation-related data structures. Detach is not safe for a thread like this.

(there's only one use where using `detach` is acceptable, that is the treads spawning child processes for notification that run free)

**Comment by fanquake** [comment]
Pushed a `.detach()`-less change.

File: src/init.cpp [hunk 2]
```diff
@@ -214,8 +216,9 @@ void Shutdown(NodeContext& node)
     StopTorControl();
 
     // After everything has been shut down, but before things get flushed, stop the
-    // CScheduler/checkqueue threadGroup
+    // CScheduler/checkqueue, threadGroup and load block thread.
     if (node.scheduler) node.scheduler->stop();
+    g_load_block.join();
```
Comment (donaloconnor) [inline src/init.cpp:221]: I think it might be safer to check if the thread is joinable before joining in case this path is reached after some initialization failure when the thread may not have started.

Comment (fanquake) [reply src/init.cpp]: Thanks, I've added that.

**Comment by donaloconnor** [comment]
ACK 83fd3a6

File: src/init.cpp [hunk 3]
```diff
@@ -152,6 +152,8 @@ NODISCARD static bool CreatePidFile()
 
 static std::unique_ptr<ECCVerifyHandle> globalVerifyHandle;
 
+static std::thread g_load_block;
```
Comment (MarcoFalke) [inline src/init.cpp:155]: I don't really like adding more and more globals to init. Ideally init should be a slim file with only calls to other modules to start/stop. Longer term `ThreadImport` should be moved to validation because it is purely validation related. Init shouldn't need to know how to load blocks from disk, etc... Note that `LoadMempool` is also in validation.

Moreover, with assumeutxo, thus multiple chainstates, I could imagine that a reindex/import can be done in the background chainstate for additional efficiency. Having init to deal with even more chainstate related logic would feel wrong.

What do you think about making this thread a member of the chainstate manager?

Comment (laanwj) [reply src/init.cpp:155]: On the other hand this was a *hidden* global before, a part of `threadGroup`. I don't think this PR makes anything worse.

I think your longer-term solution makes sense, but I'm not sure it needs to be done here.

**Comment by laanwj** [comment]
Code review ACK 83fd3a6d73eee452dc5141bdf6826da62d7b2dbd

File: src/init.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/init.cpp:155]: Agree that it doesn't make anything worse, but it could be slightly better with a trivial diff.

```diff
diff --git a/src/init.cpp b/src/init.cpp
index 8d9566edc3..984f6b2034 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -152,8 +152,6 @@ NODISCARD static bool CreatePidFile()
 
 static std::unique_ptr<ECCVerifyHandle> globalVerifyHandle;
 
-static std::thread g_load_block;
-
 static boost::thread_group threadGroup;
 
 void Interrupt(NodeContext& node)
@@ -218,7 +216,7 @@ void Shutdown(NodeContext& node)
     // After everything has been shut down, but before things get flushed, stop the
     // CScheduler/checkqueue, threadGroup and load block thread.
     if (node.scheduler) node.scheduler->stop();
-    if (g_load_block.joinable()) g_load_block.join();
+    if (node.chainman && node.chainman->m_load_block.joinable()) node.chainman->m_load_block.join();
     threadGroup.interrupt_all();
     threadGroup.join_all();
 
@@ -1844,7 +1842,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node)
         vImportFiles.push_back(strFile);
     }
 
-    g_load_block = std::thread(&TraceThread<std::function<void()>>, "loadblk", [=, &chainman]{ ThreadImport(chainman, vImportFiles); });
+    chainman.m_load_block = std::thread(&TraceThread<std::function<void()>>, "loadblk", [=, &chainman]{ ThreadImport(chainman, vImportFiles); });
 
     // Wait for genesis block to be processed
     {
diff --git a/src/validation.h b/src/validation.h
index e403bcb51a..9764a7aaca 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -783,6 +783,7 @@ private:
     friend CChain& ChainActive();
 
 public:
+    std::thread m_load_block;
     //! A single BlockManager instance is shared across each constructed
     //! chainstate to avoid duplicating block metadata.
     BlockManager m_blockman GUARDED_BY(::cs_main);

**Review by maflcko — COMMENTED** [review]
ACK 83fd3a6d73eee452dc5141bdf6826da62d7b2dbd

**Review by hebasto — APPROVED** [review]
ACK 83fd3a6d73eee452dc5141bdf6826da62d7b2dbd, I have reviewed the code and it looks OK, I agree it can be merged.

While this code is touched already, the "loadblk" thread could be renamed to "loadblocks" (10 characters from 13 allowed) or "importblocks" (12 characters) for readability.

File: src/init.cpp (hunk 3 above)
Comment (fanquake) [reply src/init.cpp:155]: Thanks. I'll take care of the suggestions, adding to chainman and thread renaming, in some followup changes. I just wanted to get it out of the Boost thread group here.
